### PR TITLE
SDIT-902 Add offender, offender booking and incentives to the new DSL

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 1]
```diff
@@ -681,12 +681,12 @@ class ActivityResourceIntTest : IntegrationTestBase() {
               courseScheduleRule()
             }
           }
-          offenderBooking = offender(nomsId = "A1234TT") {
-            booking(agencyLocationId = "LEI") {
+          newOffender(nomsId = "A1234TT") {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:684]: Created a new entry point for Offender that uses the new style of DSL builders etc. This is to allow it to work in conjunction with the existing `offender` entry point. Once all entities related to `offender` have been modelled in the new DSL we can delete it and rename `newOffender`.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt [hunk 2]
```diff
@@ -718,12 +718,12 @@ class ActivityResourceIntTest : IntegrationTestBase() {
...
+            offenderBooking = booking(agencyLocationId = "LEI") {
               incentive(iepLevelCode = "STD")
               courseAllocation(courseActivity)
             }
-          }.latestBooking()
+          }
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityResourceIntTest.kt:726]: All calls to `latestBooking()` have been removed - the booking is returned from the DSL now.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseAllocation.kt [hunk 3]
```diff
@@ -15,15 +17,15 @@ import java.time.LocalDate
...
+@NomisDataDslMarker
 interface CourseAllocationDsl {
 
   @CourseAllocationPayBandDslMarker
   fun payBand(
     startDate: String = "2022-10-31",
     endDate: String? = null,
     payBandCode: String = "5",
-  )
+  ): OffenderProgramProfilePayBand
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseAllocation.kt:28]: I was able to convert courseAllocation, courseAllocationPayBand and courseAttendance to the new DSL style as they now live underneath the `newOffender` DSL.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Incentive.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.time.LocalDateTime
+
+@DslMarker
+annotation class NewIncentiveDslMarker
+
+@NomisDataDslMarker
+interface NewIncentiveDsl
+
+@Component
+class NewIncentiveBuilderRepository(
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Incentive.kt:21]: Modelled incentives in the new DSL style. As we already have an IncentiveBuilder and DSLMarker etc. I chose to prefix everything with `New` to indicate its using the new DSL. Everything can be renamed once the old DSL has been replaced.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Offender.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  @NewBookingDslMarker
+  fun booking(
+    bookingBeginDate: LocalDateTime = LocalDateTime.now(),
+    active: Boolean = true,
+    inOutStatus: String = "IN",
+    youthAdultCode: String = "N",
+    agencyLocationId: String = "BXI",
+    dsl: NewBookingDsl.() -> Unit = {},
+  ): OffenderBooking
+}
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Offender.kt:28]: Similarly prefixed with `New` until the old DSL for offender has been replaced.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderBookingBuilder.kt [hunk 6]
```diff
@@ -98,19 +95,4 @@ class OffenderBookingBuilder(
...
-    courseActivity: CourseActivity,
-    startDate: String?,
-    programStatusCode: String,
-    endDate: String?,
-    payBands: MutableList<CourseAllocationPayBandBuilder>,
-    endReasonCode: String?,
-    endComment: String?,
-    attendances: MutableList<CourseAttendanceBuilder>,
-    dsl: CourseAllocationDsl.() -> Unit,
-  ) {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderBookingBuilder.kt:112]: Nothing is using the old style course allocation DSL so was able to remove entirely.

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**